javascript/ql/test/query-tests/Security/CWE-601/ServerSideUrlRedirect/ServerSideUrlRedirectGood2.js

```diff
@@ -0,0 +1,15 @@
+const app = require("express")();
+
+function isLocalUrl(url) {
+  return url.startsWith("/") && !url.startsWith("//") && !url.startsWith("/\\");
+}
```

Comment:
I think your idea is OK, but this sanitizer is not. 
You want to be able to redirect to `foo/bar`, but you are requiring that all urls must start with a single `/`.  

ChatGPT suggested the below, I think it's nicer. 

```JavaScript
function isRelativePath(path) {
    return !/^(\w+:)?\/\//.test(path);
}
```

Although maybe the suggested sanitizer should be generalized a bit to also handle backslash.

Reply:
Oops, yeah, that was a bit overeager. The proposed alternative looks plausible, though as mentioned above I have very limited confidence in my ability to spot ways of circumventing this sanitiser. In my local experiments, I didn't manage to provoke a cross-domain redirect with backslashes, but I suppose there's a reason the query checks for it. (I haven't changed the example to check for backslashed yet.)

Reply:
Backslash sometimes acts the same as forward slash in URLs. 
I'm not sure why that is the case, and it's something very few people seem to be aware of.

E.g. try to execute this in your JS console. 
```JavaScript
window.open("https:\\\\google.com")
```

So yeah, the above `isRelativePath ` is what ChatGPT suggested, but something like:  

```JavaScript
function isRelativePath(path) {
    return !/^(\w+:)?(\/|\\){2}/.test(path);
}
```
Is probably better.

Reply:
> Backslash sometimes acts the same as forward slash in URLs.

I realise that. I'm just saying that I wasn't able to provoke a problematic redirect with backslashes in the example code. Still, probably best to fix it.

Reply:
**[]()[]()@![30a1367d3adc27564788ec85351da896.jpg](https://github.com/github/codeql/assets/135707245/75e9ae06-8acc-43a5-84c6-51b57aecd5d6)

> # - [ ] @~~````

````~~ **```

```

Reply:
OK please help me fix it